clippy_lints/src/undocumented_unsafe_blocks.rs

```diff
@@ -0,0 +1,128 @@
+    fn gather_safety_comments(sm: &SourceMap, file: &FileName) -> Option<Vec<SafetyComment>> {
+        let source_file = sm.get_source_file(file)?;
+        let src = source_file.src.as_deref()?;
+        let start_bpos = source_file.start_pos;
+        let mut comments = Vec::new();
+
+        let mut inside_comment = false;
+        let mut pos = rustc_lexer::strip_shebang(src).unwrap_or(0);
+        for token in rustc_lexer::tokenize(&src[pos..]) {
+            match token.kind {
+                rustc_lexer::TokenKind::LineComment { doc_style: None }
+                | rustc_lexer::TokenKind::BlockComment {
+                    doc_style: None,
+                    terminated: true,
+                } if src[pos + 2..pos + token.len].trim_start().starts_with("SAFETY") => {
+                    inside_comment = true;
+                },
+                rustc_lexer::TokenKind::LineComment { doc_style: None }
+                | rustc_lexer::TokenKind::BlockComment { doc_style: None, .. }
+                | rustc_lexer::TokenKind::Whitespace => {},
+                _ => {
+                    if inside_comment {
+                        let hi = start_bpos + BytePos(pos.try_into().unwrap());
+                        comments.push(SafetyComment {
+                            hi,
+                            hi_line: source_file.lookup_file_pos_with_col_display(hi).0,
+                        });
+                        inside_comment = false;
+                    }
+                },
+            }
+            pos += token.len;
+        }
+        Some(comments)
+    }
+}
+
+impl_lint_pass!(UndocumentedUnsafeBlocks => [UNDOCUMENTED_UNSAFE_BLOCKS]);
+
+impl EarlyLintPass for UndocumentedUnsafeBlocks {
+    fn check_expr(&mut self, cx: &EarlyContext<'_>, expr: &rustc_ast::Expr) {
+        let (block, comments) = if_chain! {
+            if let ExprKind::Block(ref block, _) = expr.kind;
+            if let BlockCheckMode::Unsafe(UnsafeSource::UserProvided) = block.rules;
+            if let Some(comments) = self.safety_comments(cx.sess.source_map(), block.span);
+            then {
+                (block, comments)
+            }
+            else {
+                return;
+            }
+        };
+        if_chain! {
+            // Since we're consuming comments as we visit the AST, the comment
+            // we're searching for is likely to be at the beginning of the vector,
+            // so this is better than a binary search.
+            if let Some((i, comment)) = comments.iter().enumerate().take_while(|(_, c)| c.hi <= block.span.lo()).last();
+            let block_line = cx.sess.source_map().lookup_char_pos(block.span.lo()).line;
+            if block_line == comment.hi_line + 1 || block_line == comment.hi_line;
+            then {
+                comments.remove(i);
+            }
+            else {
+                span_lint(cx, UNDOCUMENTED_UNSAFE_BLOCKS, expr.span, "this `unsafe` block is missing a SAFETY comment");
```

Comment:
I think we should produce a help message here, suggesting to add a `// SAFETY: ...`  above the `unsafe` block.